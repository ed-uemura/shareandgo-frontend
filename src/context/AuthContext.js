import React, { createContext, useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { auth, db } from '../firebaseConfig';
import {
	createUserWithEmailAndPassword,
	signInWithEmailAndPassword,
	signOut,
	deleteUser,
} from 'firebase/auth';
import {
	doc,
	collection,
	addDoc,
	setDoc,
	getDoc,
	deleteDoc,
	updateDoc,
	query,
	where,
	getDocs,
} from 'firebase/firestore';

const Context = createContext();

const AuthContext = ({ children }) => {
	const [authenticated, setAuthenticated] = useState(false);
	const [loading, setLoading] = useState(true);
	const [userData, setUserData] = useState({});
	const [loggedUser, setLoggedUser] = useState({});
	const [destination, setDestination] = useState({});
	const [setupRide, setSetupRide] = useState({});
	const [rideData, setRideData] = useState([]);
	const [rideSelection, setRideSelection] = useState({});
	const navigate = useNavigate();

	useEffect(() => {
		if (!loggedUser) {
			setAuthenticated(false);
		} else if (loggedUser) {
			setAuthenticated(true);
		}

		if (authenticated) {
			const getUser = async () => {
				try {
					const docRef = doc(db, 'users', loggedUser.uid);
					const docSnap = await getDoc(docRef);

					if (docSnap.exists()) {
						setUserData(docSnap.data());
					} else {
						console.log('No document');
					}
				} catch (err) {
					console.log(err.message);
				}
			};

			getUser();
		}

		setLoading(false);
	}, [loggedUser, setUserData, authenticated]);

	//Handle Registration
	const handleRegistration = async (values) => {
		try {
			let userUid;
			await createUserWithEmailAndPassword(
				auth,
				values.email,
				values.password
			).then((cred) => {
				userUid = cred.user.uid;
			});

			if (values.accountType === 'Passenger') {
				await setDoc(doc(db, 'users', userUid), {
					firstName: values.firstName,
					lastName: values.lastName,
					accountType: values.accountType,
					phone: values.phone,
					address: {
						street: values.address,
						city: values.city,
						province: values.province,
						zip: values.zip,
					},
					photo: values.photo,
					payment: {
						nameOnCard: values.nameCard,
						cardNumber: values.cardNumber,
						cardExpirationDate: values.expirationDate,
						cardCsc: values.csc,
						cardZipCode: values.cardZipCode,
					},
					rides: [],
					evaluation: '5',
				})
					.then(() => {
						return navigate('/main-passenger');
					})
					.catch((err) => {
						console.log(err.message);
						handleSignOut();
						return navigate('/');
					});
			} else if (values.accountType === 'Driver') {
				await setDoc(doc(db, 'users', userUid), {
					firstName: values.firstName,
					lastName: values.lastName,
					accountType: values.accountType,
					phone: values.phone,
					address: {
						street: values.address,
						city: values.city,
						province: values.province,
						zip: values.zip,
					},
					photo: values.photo,
					payment: {
						bankName: values.bankName,
						institutionNumber: values.institutionNumber,
						transitNumber: values.transitNumber,
						accountNumber: values.accountNumber,
					},
					rides: [],
					evaluation: '5',
				})
					.then(() => {
						return navigate('/main-driver');
					})
					.catch((err) => {
						console.log(err.message);
						handleSignOut();
						return navigate('/');
					});
			}
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Login
	const handleSignIn = async (values) => {
		try {
			let userUid;
			await signInWithEmailAndPassword(auth, values.email, values.password)
				.then((resp) => {
					userUid = resp.user.uid;
				})
				.catch((err) => {
					console.log(err.message);
					handleSignOut();
					return navigate('/');
				});

			const docRef = doc(db, 'users', userUid);
			const docSnap = await getDoc(docRef);

			if (docSnap.exists()) {
				const docData = docSnap.data();

				if (docData.accountType === 'Driver') {
					return navigate('/main-driver');
				} else if (docData.accountType === 'Passenger') {
					return navigate('/main-passenger');
				} else {
					handleSignOut();
					return navigate('/');
				}
			} else {
				console.log('no data found');
				handleSignOut();
				return navigate('/');
			}
		} catch (err) {
			console.log(err.message);
			handleSignOut();
			return navigate('/');
		}
	};

	//Handle Logout
	const handleSignOut = async () => {
		await signOut(auth);
		return navigate('/');
	};

	//Handle Delete User
	const handleDeleteUser = async (event) => {
		event.preventDefault();
		try {
			await deleteDoc(doc(db, 'users', loggedUser.uid));
			deleteUser(auth.currentUser)
				.then(() => {
					handleSignOut();
					navigate('/');
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Setup Ride
	const handlesetUpRide = async (e, values) => {
		e.preventDefault();

		const rideSetup = {
			currentLocation: values.currentLocation,
			destinationLocation: values.destinationLocation,
			rideDate: values.rideDate,
			rideTime: values.rideTime,
			numOfSeats: values.numOfSeats,
			rideDetails: values.rideDetails,
			driver: {
				uid: auth.currentUser.uid,
				photo: userData.photo,
				firstName: userData.firstName,
				lastName: userData.lastName,
				rating: userData.evaluation,
			},
			passengers: [],
			passengerNum: 0,
			price: '82.40',
			reviews: [],
		};

		setSetupRide(rideSetup);

		navigate('/create-ride');
	};

	//Handle create ride
	const handleCreateRide = async (e, values) => {
		e.preventDefault();

		let rideId;

		try {
			await addDoc(collection(db, 'rides'), values)
				.then((resp) => {
					rideId = resp.id;
				})
				.catch((err) => {
					console.log(err.message);
				});

			const docRef = doc(db, 'rides', rideId);
			await updateDoc(docRef, {
				uid: rideId,
			})
				.then(() => {
					return navigate(`/ride-info/${rideId}`);
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	//Handle Cancel Ride
	const handleDestroyRide = async (id) => {
		const docRef = doc(db, 'rides', id);

		await deleteDoc(docRef)
			.then(() => {
				navigate('/main-driver');
			})
			.catch((err) => {
				console.log(err.message);
			});
	};

	//Handle Cancel Ride
	const handleFinishRide = async (id) => {
		const docRef = doc(db, 'rides', id);

		await deleteDoc(docRef)
			.then(() => {
				navigate('/ride-complete');
			})
			.catch((err) => {
				console.log(err.message);
			});
	};

	//Handle Get Rides
	const handleGetRides = async (values) => {
		const rideDataArray = [];

		const q = query(
			collection(db, 'rides'),
			where('destinationLocation', '==', values.destinationLocation)
		);

		try {
			const querySnapshot = await getDocs(q);
			querySnapshot.forEach((doc) => {
				rideDataArray.push(doc.data());
			});
		} catch (err) {
			console.log(err.message);
		}

		setRideData(rideDataArray);
		setRideSelection(values);

		navigate('/available-drivers');
	};

	// Handle Request Ride
	const handleRequestRide = async (rideId, numOfPassengers) => {
		try {
			const docRef = doc(db, 'rides', rideId);

			await updateDoc(docRef, {
				passengerNum: numOfPassengers + 1,
				passengers: [
					{
						uid: loggedUser.uid,
						firstName: userData.firstName,
						lastName: userData.lastName,
						fullName: userData.firstName + ' ' + userData.lastName,
						photo: userData.photo,
					},
				],
			})
				.then(() => {
					return navigate('/waiting-ride');
				})
				.catch((err) => {
					console.log(err.message);
				});
		} catch (err) {
			console.log(err.message);
		}
	};

	return (
		<Context.Provider
			value={{
				authenticated,
				setAuthenticated,
				loading,
				setLoading,
				userData,
				setUserData,
				loggedUser,
				setLoggedUser,
				destination,
				setDestination,
				setupRide,
				setSetupRide,
				rideData,
				setRideData,
				rideSelection,
				setRideSelection,
				handleRegistration,
				handleSignIn,
				handleSignOut,
				handleDeleteUser,
				handlesetUpRide,
				handleCreateRide,
				handleDestroyRide,
				handleFinishRide,
				handleGetRides,
				handleRequestRide,
			}}>
			{children}
		</Context.Provider>
	);
};

export { Context, AuthContext };
