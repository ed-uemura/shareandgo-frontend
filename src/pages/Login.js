import React, { useState } from 'react';
import ShareGoIcon from '../images/ShareGo-icon.svg';
import ButtonPrimary from '../components/ButtonPrimary';
import axios from 'axios';

const Login = () => {
	const [values, setValues] = useState('');
	const [error, setError] = useState("");

	function handleChange(event) {
		const auxValues = { ...values };
		auxValues[event.target.id] = event.target.value;
		setValues(auxValues);
	}

	// submit form
	const handleSubmit = async (e) => {
		e.preventDefault();
		// Fields Validation
		if(!values.email) 
			setError('Please enter an email address');
		else if(!values.password) 
			setError('Please enter a password');
		else {
			setError('');
			// Prepare user object
			const user = {
				email: values.email,
				password: values.password
			};
			try {
				// Send user object to server
				const res = await axios.post(process.env.REACT_APP_BACKEND_URL+"user/login", user);
				if(res.data.error) {
					// If error, display error message
					console.log(res);
					setError(res.data.error);
				}
				else{
					// If no error, save user data to session storage
					localStorage.setItem("userToken",res.data.token);
					localStorage.setItem("userID",res.data._id);
					localStorage.setItem("userFirstName",res.firstname);
					localStorage.setItem("userType",res.accounttype);
					values.accountType === 'driver' ? window.location.href = '/main-driver' : window.location.href = '/main-passenger';
				}
			} catch (err) {
				setError(err.response.data.error);
			}
		}
	};

	return (
		<>
			<div className='text-center'>
				<figure className='mb-5'>
					<img src={ShareGoIcon} alt='Share & Go icon' />
				</figure>
				<form className='row col-12 col-md-6 g-4 mb-4 mx-auto'>
					<div className='col-md-12'>
						<input
							type='email'
							id='email'
							placeholder='Email'
							className='form-control'
							onChange={(e) => handleChange(e)}
						/>
					</div>

					<div className='col-md-12'>
						<input
							type='password'
							id='password'
							placeholder='Password'
							className='form-control'
							onChange={(e) => handleChange(e)}
						/>
					</div>
					<div className="text-danger text-center" role="alert">
						{error}
					</div>
				</form>

				<ButtonPrimary
					text='Login'
					className='col-6 col-md-3 mx-auto d-block'
					// link='/main-passenger'
					clickAction={handleSubmit}
				/>
			</div>
		</>
	);
};

export default Login;
