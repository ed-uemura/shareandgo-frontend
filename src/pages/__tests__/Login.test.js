import {render, screen, cleanup, getByAltText} from '@testing-library/react'
import renderer from 'react-test-renderer'
import Login from '../Login'
import { MemoryRouter } from 'react-router-dom';
import * as AuthContext from '../../context/AuthContext'
import ShareGoIcon from '../../images/ShareGo-icon.svg';

afterEach(() =>{
    cleanup();
});

// test('login button should be rendered', () =>{
//     render(<MemoryRouter>
//         <Login/>
//     </MemoryRouter>)
//     const loginButtonEl = screen.getByText(/login/i);
//     expect(loginButtonEl).toBeInTheDocument();
// })

test('login button should be rendered', () =>{
    render(       
     <AuthContext.Context.Provider value={function setAuthenticated(param){}}>
        <MemoryRouter>
            <Login/>
        </MemoryRouter>
    </AuthContext.Context.Provider>)
    const iconPhotoEl = screen.getByText(/login/i);
    expect(iconPhotoEl).toBeInTheDocument();
});

test('icon should be correctly rendered', async () =>{
    
    const {getByAltText} = await render(       
        <AuthContext.Context.Provider value={function setAuthenticated(param){}}>
           <MemoryRouter>
               <Login/>
           </MemoryRouter>
       </AuthContext.Context.Provider>);

    const image = getByAltText('Share & Go icon');
    expect(image).toBeInTheDocument();
    expect(image.src).toContain(ShareGoIcon);
})

test('matches snapshot', () =>{ 
    const tree = renderer.create(
        <AuthContext.Context.Provider value={function setAuthenticated(param){}}>
        <MemoryRouter>
            <Login/>
        </MemoryRouter>
        </AuthContext.Context.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

describe('form inputs', () =>{
    test('email input should be rendered', ()=>{
        render(       
            <AuthContext.Context.Provider value={function setAuthenticated(param){}}>
            <MemoryRouter>
                <Login/>
            </MemoryRouter>
        </AuthContext.Context.Provider>)
        const usernameInputEl = screen.getByPlaceholderText(/email/i);
        expect(usernameInputEl).toBeInTheDocument();
    })

    test('password input should be rendered', ()=>{
        render(
            <AuthContext.Context.Provider value={function setAuthenticated(param){}}>
            <MemoryRouter>
                <Login/>
            </MemoryRouter>
        </AuthContext.Context.Provider>
        )
        const passwordInputEl = screen.getByPlaceholderText(/password/i);
        expect(passwordInputEl).toBeInTheDocument();

    })
})