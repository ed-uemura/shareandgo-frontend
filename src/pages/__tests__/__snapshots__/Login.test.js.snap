// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="text-center"
>
  <figure
    className="mb-5"
  >
    <img
      alt="Share & Go icon"
      src="ShareGo-icon.svg"
    />
  </figure>
  <form
    className="row col-12 col-md-6 g-4 mb-4 mx-auto"
  >
    <div
      className="col-md-12"
    >
      <input
        className="form-control"
        id="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
      />
    </div>
    <div
      className="col-md-12"
    >
      <input
        className="form-control"
        id="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
    </div>
    <div
      className="text-danger text-center"
      role="alert"
    />
  </form>
  <a
    className="btn btn-primary col-6 col-md-3 mx-auto d-block"
    href="/"
    onClick={[Function]}
  >
    Login
  </a>
</div>
`;
